python/cudf/cudf/utils/ioutils.py

```diff
@@ -1391,3 +1508,182 @@ def _ensure_filesystem(passed_filesystem, path):
+    if not byte_ranges:
+        # Early return
+        return new_ranges
+
+    offset, size = byte_ranges[0]
+    for (new_offset, new_size) in byte_ranges[1:]:
+        gap = new_offset - (offset + size)
+        if gap > max_gap or (size + new_size + gap) > max_block:
+            # Gap is too large or total read is too large
+            new_ranges.append((offset, size))
+            offset = new_offset
+            size = new_size
+            continue
+        size += new_size + gap
+    new_ranges.append((offset, size))
+    return new_ranges
+
+
+def _assign_block(fs, path_or_fob, local_buffer, offset, nbytes):
+    if fs is None:
+        # We have an open fsspec file object
+        path_or_fob.seek(offset)
+        local_buffer[offset : offset + nbytes] = np.frombuffer(
+            path_or_fob.read(nbytes), dtype="b",
+        )
+    else:
+        # We have an fsspec filesystem and a path
+        with fs.open(path_or_fob, mode="rb", cache_type="none") as fob:
+            fob.seek(offset)
+            local_buffer[offset : offset + nbytes] = np.frombuffer(
+                fob.read(nbytes), dtype="b",
+            )
+
+
+class ReadBlockWorker(Thread):
+    def __init__(self, queue, fs, path_or_fob, local_buffer):
+        Thread.__init__(self)
+        self.queue = queue
+        self.fs = fs
+        self.path_or_fob = path_or_fob
+        self.local_buffer = local_buffer
+
+    def run(self):
+        while True:
+            # Get the work from the queue and expand the tuple
+            offset, nbytes = self.queue.get()
+            try:
+                _assign_block(
+                    self.fs,
+                    self.path_or_fob,
+                    self.local_buffer,
+                    offset,
+                    nbytes,
+                )
+            finally:
+                self.queue.task_done()
+
+
+def _read_byte_ranges(
+    path_or_fob, ranges, local_buffer, fs=None, num_threads=None, **kwargs,
+):
+
+    # If we have a fs object that supports `cat_ranges`,
+    # the ranges should be collected concurrently
+    if fs is not None and hasattr(fs, "cat_ranges"):
+        paths, starts, ends = [], [], []
+        for offset, nbytes in ranges:
+            paths.append(path_or_fob)
+            starts.append(offset)
+            ends.append(offset + nbytes)
+        blocks = fs.cat_ranges(paths, starts, ends)
+        for block, (offset, nbytes) in zip(blocks, ranges):
+            local_buffer[offset : offset + nbytes] = np.frombuffer(
+                block, dtype="b",
+            )
+        return
+
+    # No reason to generate more threads than byte-ranges
+    if num_threads is None:
+        num_threads = len(ranges)
+    num_threads = min(num_threads, len(ranges))
+
+    if num_threads > 1:
+        queue = Queue()
+        for x in range(num_threads):
+            worker = ReadBlockWorker(queue, fs, path_or_fob, local_buffer)
+            worker.daemon = True
+            worker.start()
```

Comment:
Note that we **cannot** use daemon threads with a queue like this, because the thread will stay alive for the duration of the workers life.  While the thread is alive, it will continute to hold a reference to the dummy array (preventing garbage collection of the host memory buffer after IO is completed).  [nvt#1119](https://github.com/NVIDIA/NVTabular/pull/1119) uses a simpler threading approach to avoid this problem.